nibabel/parrec.py

```diff
@@ -1008,13 +1018,71 @@ def get_sorted_slice_indices(self):
             array, and (equivalently) the only dimension of
             ``self.image_defs``.
         """
-        slice_nos = self.image_defs['slice number']
-        is_full = vol_is_full(slice_nos, self.general_info['max_slices'])
-        keys = (slice_nos, vol_numbers(slice_nos), np.logical_not(is_full))
-        # Figure out how many we need to remove from the end, and trim them
-        # Based on our sorting, they should always be last
+        if not self.strict_sort:
+            slice_nos = self.image_defs['slice number']
+            is_full = vol_is_full(slice_nos, self.general_info['max_slices'])
+            keys = (slice_nos, vol_numbers(slice_nos), np.logical_not(is_full))
+            sort_order = np.lexsort(keys)
+        else:
+            # Sort based on a larger number of keys.  This is more complicated
+            # but works for .PAR files that get missorted by the above method
+            slice_nos = self.image_defs['slice number']
+            dynamics = self.image_defs['dynamic scan number']
+            phases = self.image_defs['cardiac phase number']
+            echos = self.image_defs['echo number']
+
+            # try adding keys only present in a subset of .PAR files
+            try:
+                # only present in PAR v4.2+
+                asl_labels = self.image_defs['label type']
+                asl_keys = (asl_labels, )
+            except:
+                asl_keys = ()
+            if not self.general_info['diffusion'] == 0:
+                try:
+                    # only present for .PAR v4.1+
+                    bvals = self.image_defs['diffusion b value number']
+                    bvecs = self.image_defs['gradient orientation number']
+                except:
+                    bvals = self.image_defs['diffusion_b_factor']
+                    # use hash to get a single sortable value per direction
+                    bvecs = [hash(tuple(
```

Comment:
I didn't have luck with tuple of tuples in lexsort, so I think I will handle this via:

```python
bvecs = _direction_numbers(self.get_def('diffusion'))
```
where I define the utility function
```python
def _direction_numbers(bvecs):
    """ Enumerate directions from an (N, 3) array of direction vectors.
    """
    cnt = 0
    bvec_dict = {}
    bvec_nos = np.zeros(len(bvecs))
    for i, bvec in enumerate(bvecs):
        bv = tuple(bvec)
        if bv not in bvec_dict:
            bvec_dict[bv] = cnt
            cnt += 1
        bvec_nos[i] = bvec_dict.get(bv)
    return bvec_nos
```